File: ios/Classes/THEOliveView.swift
//
//  THEOliveView.swift
//  flutter_theolive
//
//  Created by Daniel on 16/07/2023.
//

import Foundation
import Flutter
import UIKit
import THEOliveSDK
import os

let log = OSLog(subsystem: Bundle.main.bundleIdentifier ?? "live.theo.THEOlive.Flutter" , category: "THEOliveView")

class THEOliveView: NSObject, FlutterPlatformView, THEOlivePlayerEventListener, THEOliveNativeAPI {

    private static var TAG = "FL_IOS_THEOliveView"
    private var _view: UIView
    private let _viewId: Int64
    private let _flutterAPI: THEOliveFlutterAPI
    private let _pigeonMessenger: PigeonMultiInstanceBinaryMessengerWrapper

    private let player = THEOliveSDK.THEOlivePlayer()
    
    var newPlayerView: THEOliveSDK.THEOliveChromelessPlayerView?
    
    private let emptyCompletion: () -> Void = {}
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger?
    ) {
        _viewId = viewId
        _view = UIView()
        _view.frame = frame

        //setup pigeon
        _pigeonMessenger = PigeonMultiInstanceBinaryMessengerWrapper(with: messenger!, channelSuffix: "id_\(viewId)")
        _flutterAPI = THEOliveFlutterAPI(binaryMessenger: _pigeonMessenger)

        super.init()

        THEOliveNativeAPISetup.setUp(binaryMessenger: _pigeonMessenger, api: self)

        // iOS views can be created here
        createNativeView(view: _view)
        setupEventListeners()

    }


    func view() -> UIView {
        return _view
    }

    func createNativeView(view _view: UIView){
        _view.backgroundColor = UIColor.black

        let newPlayerView = THEOliveSDK.THEOliveChromelessPlayerView(player: player)

        newPlayerView.translatesAutoresizingMaskIntoConstraints = false
        newPlayerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]

        newPlayerView.frame = _view.bounds

        _view.addSubview(newPlayerView)

        self.newPlayerView = newPlayerView

    }

    func setupEventListeners() {

        player.add(eventListener: self)

    }

    // THEOliveNativeAPI
    func loadChannel(channelID: String) throws {
        os_log("loadChannel: %@", log: log, type: .debug, channelID)
        self.player.loadChannel(channelID)
    }

    func play() throws {
        self.player.play()
    }

    func pause() throws {
        self.player.pause()
    }

    func preloadChannels(channelIDs: [String]) throws {
        self.player.preloadChannels(channelIDs)
    }


    // Fix for https://github.com/flutter/flutter/issues/97499
    // The PlatformViews are not deallocated in time, so we clean up upfront.
    func manualDispose() throws {
        os_log("manualDispose", log: log, type: .debug)
        player.remove(eventListener: self)
        newPlayerView?.removeFromSuperview()
        player.reset()

    }

    // THEOlivePlayerEventListener
    func onChannelLoaded(channelId: String) {
        os_log("onChannelLoaded: %@", log: log, type: .debug, channelId)
        _flutterAPI.onChannelLoadedEvent(channelID: channelId, completion: emptyCompletion)
    }
    
    
    func goLive() throws {
        os_log("onPlaying", log: log, type: .debug)
        player.goLive()
    }

    func onPlaying() {
        os_log("onPlaying", log: log, type: .debug)
        _flutterAPI.onPlaying(completion: emptyCompletion);
    }

    func onError(message: String) {
        os_log("onError: %@" , log: log, type: .debug, message)
        _flutterAPI.onError(message: message, completion: emptyCompletion)
    }

    func onChannelOffline(channelId: String) {
        os_log("onChannelOffline: %@" , log: log, type: .debug, channelId)
        _flutterAPI.onChannelOfflineEvent(channelID: channelId, completion: emptyCompletion)
    }

    func onChannelLoadStart(channelId: String) {
        os_log("onChannelLoadStart: %@" , log: log, type: .debug, channelId)
        _flutterAPI.onChannelLoadStartEvent(channelID: channelId, completion: emptyCompletion)
    }

    func onWaiting() {
        os_log("onWaiting", log: log, type: .debug)
        _flutterAPI.onWaiting(completion: emptyCompletion);
    }

    func onPlay() {
        os_log("onPlay", log: log, type: .debug)
        _flutterAPI.onPlay(completion: emptyCompletion);
    }

    func onPause() {
        os_log("onPause", log: log, type: .debug)
        _flutterAPI.onPause(completion: emptyCompletion)
    }

    func onIntentToFallback() {
        os_log("onIntentToFallback", log: log, type: .debug)
        _flutterAPI.onIntentToFallback(completion: emptyCompletion)
    }

    deinit {
        os_log("deinit %d", log: log, type: .debug, _viewId)
    }
}

File: flutter_theolive_sdk_android/android/src/main/kotlin/live/theo/sdk/flutter_theolive/THEOliveView.kt
package live.theo.sdk.flutter_theolive

import android.content.Context
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import com.theolive.player.EventListener
import com.theolive.player.PlayerView
import com.theolive.player.RenderingTarget
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import live.theo.sdk.flutter_theolive.pigeon.PigeonNativePlayerConfiguration
import live.theo.sdk.flutter_theolive.pigeon.THEOliveFlutterAPI
import live.theo.sdk.flutter_theolive.pigeon.THEOliveNativeAPI



class THEOliveView(context: Context, viewId: Int, args: Any?, messenger: BinaryMessenger) : PlatformView,
    EventListener, THEOliveNativeAPI {

    private val RENDERINGTARGET_SURFACE_VIEW = "surfaceView"

    private var playerView: PlayerView

    private val constraintLayout: LinearLayout

    private val flutterApi: THEOliveFlutterAPI

    private val pigeonMessenger: PigeonMultiInstanceBinaryMessengerWrapper

    private val id : Int

    private val emptyCallback: () -> Unit = {}

    private val nativeRenderingTarget: String
    // Workaround to eliminate the inital transparent layout with initExpensiveAndroidView
    // TODO: remove it once initExpensiveAndroidView is not used.
    private var useHybridComposition: Boolean = false
    private var isFirstPlaying: Boolean = false
        set(value) {
            if (!useHybridComposition) {
                return
            }

            if (value) {
                playerView.visibility = View.VISIBLE
            } else {
                playerView.visibility = View.INVISIBLE
            }
            field = value
        }

    // sync behavior with native iOS SDK
    // TODO: this logic should move to application level. The SDK should always stay paused after a "background to foreground" switch.
    private var wasPlayingBeforeActivityOnPause: Boolean = false;

    init {
        Log.d("THEOliveView_$viewId", "init $viewId");

        useHybridComposition = (args as? Map<*, *>)?.get("useHybridComposition") as? Boolean == true
        nativeRenderingTarget = ((args as? Map<*, *>)?.get("nativeRenderingTarget") as? String) ?: RENDERINGTARGET_SURFACE_VIEW //TODO: use enum and pigeon

        //setup pigeon
        pigeonMessenger = PigeonMultiInstanceBinaryMessengerWrapper(messenger, "id_$viewId");
        THEOliveNativeAPI.setUp(pigeonMessenger, this)
        flutterApi = THEOliveFlutterAPI(pigeonMessenger)

        constraintLayout = LinearLayout(context)

        val layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT)
        constraintLayout.layoutParams = layoutParams
        //constraintLayout.setBackgroundColor(android.graphics.Color.BLUE)

        id = viewId
        constraintLayout.id = viewId
        playerView = PlayerView(context)
        playerView.id = View.generateViewId()
        playerView.layoutParams = layoutParams
        //playerView.setBackgroundColor(android.graphics.Color.RED)

        if (nativeRenderingTarget == RENDERINGTARGET_SURFACE_VIEW) {
            Log.d("THEOliveView_$id", "nativeRenderingTarget: surfaceView");
            playerView.setRenderingTarget(RenderingTarget.SURFACE_VIEW);
        } else {
            Log.d("THEOliveView_$id", "nativeRenderingTarget: textureView");
            playerView.setRenderingTarget(RenderingTarget.TEXTURE_VIEW);
        }

        constraintLayout.addView(playerView)


        playerView.player.addEventListener(this)

    }

    override fun onChannelLoadStart(channelId: String) {
        Log.d("THEOliveView_$id", "onChannelLoadStart: $channelId");
        isFirstPlaying = false
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onChannelLoadStartEvent(channelId, emptyCallback);
        }

    }
    override fun onChannelLoaded(channelId: String) {
        Log.d("THEOliveView_$id", "onChannelLoaded:");
        isFirstPlaying = false
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onChannelLoadedEvent(channelId, emptyCallback);
        }
    }

    override fun onChannelOffline(channelId: String) {
        Log.d("THEOliveView_$id", "onChannelOffline: $channelId");
        isFirstPlaying = false
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onChannelOfflineEvent(channelId, emptyCallback);
        }
    }

    override fun onPlaying() {
        Log.d("THEOliveView_$id", "onPlaying");

        if (!isFirstPlaying) {
            isFirstPlaying = true
        }

        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onPlaying(emptyCallback)
        }
    }

    override fun onWaiting() {
        Log.d("THEOliveView_$id", "onWaiting");
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onWaiting(emptyCallback)
        }
    }

    override fun onPause() {
        Log.d("THEOliveView_$id", "onPause");
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onPause(emptyCallback)
        }
    }

    override fun onPlay() {
        Log.d("THEOliveView_$id", "onPlay");
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onPlay(emptyCallback)
        }
    }

    override fun onIntentToFallback() {
        Log.d("THEOliveView_$id", "onIntentToFallback");
        isFirstPlaying = false
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onIntentToFallback(emptyCallback)
        }
    }

    override fun onReset() {
        Log.d("THEOliveView_$id", "onReset");
        isFirstPlaying = false
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onReset(emptyCallback)
        }
    }

    override fun onError(message: String) {
        Log.d("THEOliveView_$id", "error: $message");
        isFirstPlaying = false
        CoroutineScope(Dispatchers.Main).launch {
            flutterApi.onError(message, emptyCallback)
        }
    }

    override fun getView(): View? {
        return constraintLayout
    }

    override fun dispose() {
        Log.d("THEOliveView_$id", "dispose");

        playerView.player.removeEventListener(this);
        constraintLayout.removeView(playerView)
        playerView.player.destroy()
        playerView.onDestroy()
    }

    override fun loadChannel(channelID: String) {
        Log.d("THEOliveView_$id", "loadChannel called, $channelID");
        playerView.player.loadChannel(channelID);
    }

    override fun manualDispose() {
        Log.d("THEOliveView_$id", "manualDispose");
        //DO NOTHING, normal dispose() flow should be called by Flutter
    }

    override fun onLifecycleResume() {
        Log.d("THEOliveView_$id", "onLifecycleResume");
        playerView.onResume();
        if (wasPlayingBeforeActivityOnPause) {
            wasPlayingBeforeActivityOnPause = false;
            play();
        }
    }

    override fun onLifecyclePause() {
        Log.d("THEOliveView_$id", "onLifecyclePause");
        wasPlayingBeforeActivityOnPause = !playerView.player.paused
        playerView.onPause();
    }


    override fun preloadChannels(channelIDs: List<String>) {
        this.playerView.player.preloadChannels(channelIDs.toTypedArray());
    }

    override fun play() {
        this.playerView.player.play();
    }

    override fun pause() {
        this.playerView.player.pause()
    }

    override fun goLive() {
        this.playerView.player.goLive()
    }

    override fun updateConfiguration(configuration: PigeonNativePlayerConfiguration) {
        this.playerView.player.updateConfig(configuration.sessionId)
    }

}